App Configuration: prepare release (#17240)
Azure/azure-sdk-for-go: sdk/appconfig/azappconfig/README.md

# Azure App Configuration client library for Go
Azure App Configuration is a managed service that helps developers centralize their application and feature settings simply and securely.
It allows you to create and manage application configuration settings and retrieve their revisions from a specific point in time.

[Source code][appconfig_client_src] | [Package (pkg.go.dev)][goget_azappconfig] | [Product documentation][appconfig_docs]

## Getting started
### Install packages
Install [azappconfig][goget_azappconfig] with `go get`:
```Bash
go get github.com/Azure/azure-sdk-for-go/sdk/appconfig/azappconfig
```

### Prerequisites
* An [Azure subscription][azure_sub]
* Go version 1.16 or higher

> * A Key Vault. If you need to create one, you can use the [Azure Cloud Shell][azure_cloud_shell] to create one with these commands (replace `"my-resource-group"` and `"my-app-config"` with your own, unique

Reviewer: ```suggestion
* An App Configuration store. If you need to create one, you can use the [Azure Cloud Shell][azure_cloud_shell] to create one with these commands (replace `"my-resource-group"` and `"my-app-config"` with your own, unique
```

(Unless you're trying to show the keyvault integration, perhaps?)

names):

  (Optional) if you want a new resource group to hold the Azure App Configuration:
  ```sh
  az group create --name my-resource-group --location westus2
  ```

  Create the Key Vault:
  ```Bash
  az appconfig create --resource-group my-resource-group --name my-app-config --location westus2
  ```

  Output:
  ```json
  {
      "creationDate": "...",
      "endpoint": "https://my-app-config.azconfig.io",
      "id": "/subscriptions/.../resourceGroups/my-resource-group/providers/Microsoft.AppConfiguration/configurationStores/my-app-config",
      "location": "westus2",
      "name": "my-app-config",
      "provisioningState": "Succeeded",
      "resourceGroup": "my-resource-group",
      "tags": {},
      "type": "Microsoft.AppConfiguration/configurationStores"
  }
  ```

  > The `"endpoint"` property is the `endpointUrl` used by [Client][appconfig_client_src]

### Authenticate the client
This document demonstrates using the connection string. However, [Client][appconfig_client_src] accepts any [azidentity][azure_identity] credential. See the [azidentity][azure_identity] documentation for more information about other credentials.

#### Create a client

> Constructing the client requires your vault's connection string, which you can get from the Azure Portal.

Reviewer: ```suggestion
Constructing the client requires your App Configuration connection string, which you can get from the Azure Portal.
```

Author: Thank you!

```

## Examples
This section contains code snippets covering common tasks:
* [Add a configuration setting](#add-a-configuration-setting "Add a configuration setting")
* [Get a configuration setting](#get-a-configuration-setting "Get a configuration setting")
* [Set a configuration setting](#set-a-configuration-setting "Set a configuration setting")
* [Set a configuration setting read only](#set-a-configuration-setting-read-only "Set a configuration setting read only")
* [List configuration setting revisions](#list-configuration-setting-revisions "List configuration setting revisions")
* [Delete a configuration setting](#set-a-configuration-setting "Delete a configuration setting")

### Add a configuration setting

```go
import (
    "fmt"
    "os"

    "github.com/Azure/azure-sdk-for-go/sdk/appconfig/azappconfig"
    "github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
)

func ExampleAddConfigurationSetting() {
    connectionString := os.Getenv("APPCONFIGURATION_CONNECTION_STRING")
    client, err = azappconfig.NewClientFromConnectionString(connectionString, nil)
    if err != nil {
        panic(err)
    }

    // Create configuration setting

>     resp, err := client.AddSetting(context.TODO(), azappconfig.Setting{Key: to.StringPtr("key"), Label: to.StringPtr("label"), Value: to.StringPtr("value")}, nil)

Reviewer: It would be nice to add some whitespace here, just to make it more clear you're setting three fields in a Setting:
(whatever gofmt ends up doing)

    resp, err := client.AddSetting(context.TODO(), 
      azappconfig.Setting{
        Key: to.StringPtr("key"), 
        Label: to.StringPtr("label"), 
        Value: to.StringPtr("value")
      }, nil)

Reviewer: (there are a few spots like this, in case you decide you want to change it)

Author: Thank you, I updated the samples to include more spacing/better formatting using newlines.

    if err != nil {
        panic(err)
    }
    fmt.Println(*resp.Key)
    fmt.Println(*resp.Label)
    fmt.Println(*resp.Value)
}
```

### Get a configuration setting

```go
import (
    "fmt"
    "os"

    "github.com/Azure/azure-sdk-for-go/sdk/appconfig/azappconfig"
    "github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
)

func ExampleGetConfigurationSetting() {
    connectionString := os.Getenv("APPCONFIGURATION_CONNECTION_STRING")
    client, err = azappconfig.NewClientFromConnectionString(connectionString, nil)
    if err != nil {
        panic(err)
    }

    // Get configuration setting

>     resp, err := client.GetSetting(context.TODO(), azappconfig.Setting{Key: to.StringPtr("key"), Label: to.StringPtr("label")}, nil)

Reviewer: Does 'key' need to be defined as a string pointer? I wouldn't think it'd be optional at all (label is, however).

Author: Yes - because `Setting` is a return type, and you may ask for the list of items via `ListRevisions`, and ask it to not include `Key` column. In that case, the service will return `Setting` with `nil` `Key`.

Reviewer: But this argument is for getting a setting - it has different semantics than ListSetting().

Is there a reason we're not just using 'key' as a positional argument?

Author: Then the key will be duplicated in the `Setting`, and as a parameter, I think it will be confusing.
If you look at the Java or .NET SDK, they have either methods with this semantics, taking `Setting`, or an overload with string key and optional string label, but not a mix.

Reviewer: (this isn't a "delay this preview" thing. But let's discuss it for the next one)

Reviewer: (if you don't mind filing an issue)

Author: As of ideas, initially, I had `Setting` being a struct with private fields, which you could only create as `NewSetting()` with string key parameter. It had getters to get fields as pointers, and "Setters", such as `WithLabel()`, `WithValue()` also taking strings and not pointers. Not all fields were writeable - the same as in other SDKs. And it was not possible then for user to create a setting without key. `With...()` were retutning a copy of the struct, with the corresponding field being set. But I was told such API is not Go-like, where everything is much simpler, uses bare structs, pointers, relies on comments etc.

The issue is: #17264

# Azure App Configuration client library for Go
Azure App Configuration is a managed service that helps developers centralize their application and feature settings simply and securely.
It allows you to create and manage application configuration settings and retrieve their revisions from a specific point in time.

[Source code][appconfig_client_src] | [Package (pkg.go.dev)][goget_azappconfig] | [Product documentation][appconfig_docs]

## Getting started
### Install packages
Install [azappconfig][goget_azappconfig] with `go get`:
```Bash
go get github.com/Azure/azure-sdk-for-go/sdk/appconfig/azappconfig
```

### Prerequisites
* An [Azure subscription][azure_sub]
* Go version 1.16 or higher
* A Key Vault. If you need to create one, you can use the [Azure Cloud Shell][azure_cloud_shell] to create one with these commands (replace `"my-resource-group"` and `"my-app-config"` with your own, unique
names):

  (Optional) if you want a new resource group to hold the Azure App Configuration:
  ```sh
  az group create --name my-resource-group --location westus2
  ```

  Create the Key Vault:
  ```Bash

>   az appconfig create --resource-group my-resource-group --name my-app-config --location westus2

Author: Yes, I was able to create it.

```

### Set a configuration setting

```go
import (
    "fmt"
    "os"

    "github.com/Azure/azure-sdk-for-go/sdk/appconfig/azappconfig"
)

func ExampleSetConfigurationSetting() {
    connectionString := os.Getenv("APPCONFIGURATION_CONNECTION_STRING")
    client, err = azappconfig.NewClientFromConnectionString(connectionString, nil)
    if err != nil {
        panic(err)
    }

    // Set configuration setting
    key := "key"
    label := "label"
    value := "new_value"
    resp, err := client.SetSetting(context.TODO(), Setting{Key: &key, Label: &label, Value: &value}, nil)
    if err != nil {
        panic(err)
    }
    fmt.Println(*resp.Key)
    fmt.Println(*resp.Label)
    fmt.Println(*resp.Value)
}
```

### Set a configuration setting read only

```go
import (
    "fmt"
    "os"

    "github.com/Azure/azure-sdk-for-go/sdk/appconfig/azappconfig"
)

func ExampleSetConfigurationSettingReadOnly() {
    connectionString := os.Getenv("APPCONFIGURATION_CONNECTION_STRING")
    client, err = azappconfig.NewClientFromConnectionString(connectionString, nil)
    if err != nil {
        panic(err)
    }

    // Set configuration setting read only
    key := "key"
    label := "label"

>     resp, err := client.SetReadOnly(context.TODO(), Setting{Key: &key, Label: &label, true, nil)

Reviewer 2: Also missing a bracket

    if err != nil {
        panic(err)
    }
    fmt.Println(*resp.Key)
    fmt.Println(*resp.Label)
    fmt.Println(*resp.Value)
    fmt.Println(*resp.IsReadOnly)

    // Remove read only status
    resp, err := client.SetReadOnly(context.TODO(), Setting{Key: &key, Label: &label, false, nil)
    if err != nil {
        panic(err)
    }
    fmt.Println(*resp.Key)
    fmt.Println(*resp.Label)
    fmt.Println(*resp.Value)
    fmt.Println(*resp.IsReadOnly)
}
```

### List configuration setting revisions

```go
import (
    "fmt"
    "os"

    "github.com/Azure/azure-sdk-for-go/sdk/appconfig/azappconfig"
)

func ExampleListRevisions() {
    connectionString := os.Getenv("APPCONFIGURATION_CONNECTION_STRING")
    client, err = azappconfig.NewClientFromConnectionString(connectionString, nil)
    if err != nil {
        panic(err)
    }

    any := "*"
    revPgr := client.ListRevisions(SettingSelector{KeyFilter: &any, LabelFilter: &any, Fields: AllSettingFields()}, nil)
    for revPgr.NextPage(context.TODO()) {

>         resp := revPgr.PageResponse()

Author: Which package is using the new API so I can use it as example? I'd change.

>         resp := revPgr.PageResponse()

Author: I modernized the pager. But I think in the long term we should update codegen, so that it is easier to implement and so that it does not require to expose internal members.

      "tags": {},
      "type": "Microsoft.AppConfiguration/configurationStores"
  }
  ```

  > The `"endpoint"` property is the `endpointUrl` used by [Client][appconfig_client_src]

### Authenticate the client
This document demonstrates using the connection string. However, [Client][appconfig_client_src] accepts any [azidentity][azure_identity] credential. See the [azidentity][azure_identity] documentation for more information about other credentials.

#### Create a client
Constructing the client requires your App Configuration connection string, which you can get from the Azure Portal.
```Bash
export APPCONFIGURATION_CONNECTION_STRING="Endpoint=https://my-app-config.azconfig.io;Id=...;Secret=..."
```

```go
import (
    "os"
    "github.com/Azure/azure-sdk-for-go/sdk/appconfig/azappconfig"
)

client, err = azappconfig.NewClientFromConnectionString(os.Getenv("APPCONFIGURATION_CONNECTION_STRING"), nil)
```

Or, using Default Azure Credential from Azure Identity:

```go
import (
    "github.com/Azure/azure-sdk-for-go/sdk/appconfig/azappconfig"
    "github.com/Azure/azure-sdk-for-go/sdk/azidentity"
)

credential, err := azidentity.NewDefaultAzureCredential(nil)

> client, err = azappconfig.NewClient("https://my-app-config.azconfig.io", credential, nil)

Reviewer 2: ```suggestion
client, err := azappconfig.NewClient("https://my-app-config.azconfig.io", credential, nil)
```

Reviewer 2: This occurs throughout, not going to point out all of them

    fmt.Println(*resp.Label)
    fmt.Println(*resp.Value)
    fmt.Println(*resp.IsReadOnly)

    // Remove read only status
    resp, err := client.SetReadOnly(
        context.TODO(),
        azappconfig.Setting{
            Key: to.StringPtr("key"),
            Label: to.StringPtr("label")
        },
        false,
        nil)

    if err != nil {
        panic(err)
    }

    fmt.Println(*resp.Key)
    fmt.Println(*resp.Label)
    fmt.Println(*resp.Value)
    fmt.Println(*resp.IsReadOnly)
}
```

### List configuration setting revisions

```go
import (
    "fmt"
    "os"

    "github.com/Azure/azure-sdk-for-go/sdk/appconfig/azappconfig"
    "github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
)

func ExampleListRevisions() {
    connectionString := os.Getenv("APPCONFIGURATION_CONNECTION_STRING")
    client, err = azappconfig.NewClientFromConnectionString(connectionString, nil)
    if err != nil {
        panic(err)
    }

    revPgr := client.ListRevisions(
        azappconfig.SettingSelector{
            KeyFilter: to.StringPtr("*"),
            LabelFilter: to.StringPtr("*"),
            Fields: azappconfig.AllSettingFields()
        },
        nil)

    for revPgr.More() {

>         if revResp, revErr := revPgr.NextPage(context.TODO()); revErr != nil {

Reviewer 2: ```suggestion
        if revResp, revErr := revPgr.NextPage(context.TODO()); revErr == nil {
```
